test/multiple-user-principal-types.test.js

```diff
@@ -0,0 +1,978 @@
+      User.settings.caseSensitiveEmail = false;
+      User.login(validMixedCaseEmailCredentials, function(err, accessToken) {
+        assertGoodToken(accessToken, validCredentialsUser);
+
+        done();
+      });
+    });
+
+    it('Try to login with invalid email case', function(done) {
+      User.login(validMixedCaseEmailCredentials, function(err, accessToken) {
+        assert(err);
+
+        done();
+      });
+    });
+
+    it('Login a user by providing credentials with TTL', function(done) {
+      User.login(validCredentialsWithTTL, function(err, accessToken) {
+        assertGoodToken(accessToken, validCredentialsUser);
+        assert.equal(accessToken.ttl, validCredentialsWithTTL.ttl);
+
+        done();
+      });
+    });
+
+    it('honors default `createAccessToken` implementation', function(done) {
+      User.login(validCredentialsWithTTL, function(err, accessToken) {
+        assert(accessToken.userId);
+        assert(accessToken.id);
+        assert(accessToken.principalType);
+
+        User.findById(accessToken.userId, function(err, user) {
+          user.createAccessToken(120, function(err, accessToken) {
+            assertGoodToken(accessToken, validCredentialsUser);
+            assert.equal(accessToken.ttl, 120);
+
+            done();
+          });
+        });
+      });
+    });
+
+    it('honors default `createAccessToken` implementation - promise variant', function(done) {
+      User.login(validCredentialsWithTTL, function(err, accessToken) {
+        assert(accessToken.userId);
+        assert(accessToken.id);
+        assert(accessToken.principalType);
+
+        User.findById(accessToken.userId, function(err, user) {
+          user.createAccessToken(120)
+            .then(function(accessToken) {
+              assertGoodToken(accessToken, validCredentialsUser);
+              assert.equal(accessToken.ttl, 120);
+
+              done();
+            })
+            .catch(function(err) {
+              done(err);
+            });
+        });
+      });
+    });
+
+    it('Login a user using a custom createAccessToken', function(done) {
+      var createToken = User.prototype.createAccessToken; // Save the original method
+      // Override createAccessToken
+      User.prototype.createAccessToken = function(ttl, cb) {
+        // Reduce the ttl by half for testing purpose
+        this.accessTokens.create({ttl: ttl / 2}, cb);
+      };
+      User.login(validCredentialsWithTTL, function(err, accessToken) {
+        assertGoodToken(accessToken, validCredentialsUser);
+        assert.equal(accessToken.ttl, 1800);
+
+        User.findById(accessToken.userId, function(err, user) {
+          user.createAccessToken(120, function(err, accessToken) {
+            assertGoodToken(accessToken, validCredentialsUser);
+            assert.equal(accessToken.ttl, 60);
+            // Restore create access token
+            User.prototype.createAccessToken = createToken;
```

Comment:
Uh oh, while I understand this probably comes from the original test in `test/user.test.js`, it is horribly broken. When any of the assertions above fails, then `createAccessToken` method is never restored.

The correct solution is to create a new one-of subclass of the User model where we are free to customise `createAccessToken` method as we wish.

OTOH, is this test really needed for polymorphic user relations?